Extremely slow build times for `gdscript_vm.cpp` with MSVC and optimizations enabled

### Godot version

4.2.dev (4714e95896c8db02616ea6ec7f9aff92dec1cae4)

### System information

Windows 10.0.19045 - Vulkan (Forward+) - dedicated Radeon RX Vega (Advanced Micro Devices, Inc.; 31.0.21023.2010) - AMD Ryzen 7 2700 Eight-Core Processor (16 Threads)

### Issue description

MSVC (tested with VS 2022, but I believe VS 2019 behaves the same) takes an inordinate amount of time to compile two files in the current codebase, when optimizations are enabled:
- `core/variant/variant_call.cpp`
- `modules/gdscript/gdscript_vm.cpp`

This issue is about `gdscript_vm.cpp`, I'll open another one to track `variant_call.cpp` as the resolution for each may not be the same.

On my system, with VS 2022 17.6.5 (MSVC 14.36.32532), compiling `gdscript_vm.cpp` with the default `/O2` optimization level (used when compiling without specifying a different `optimization` or `dev_build` value) takes more than 25 min (!). This is more than half of the total build time for the whole Godot codebase.

We need to figure out why MSVC struggles so much with that file, and how to improve this. It's very likely that we can at least gain an order of magnitude here - at least that should be the goal, because those build times are not acceptable (I'm on Linux 99% of the time, I don't know how MSVC users deal with this on a daily basis...).

The GDScript VM is known to be a pretty heavy file with a massive switch for all opcodes. GCC/Clang do a good job at optimizing it, so good that it's one of the reasons why our official Windows builds are made with MinGW-GCC - it makes the GDScript VM 20% faster. I believe @RandomShaper has been looking into ways to get a similar performance gain with MSVC, using some undocumented MSVC builtins. This is about runtime performance, but if we're lucky this _may_ also help with build time.

This can be debugged by compiling only the `gdscript_vm.cpp` file on the command line with `cl.exe`. You can do a build with `scons p=windows target=editor verbose=yes` to see what the command line would be, abort the build, remove the `gdscript_vm.*.obj` file, and run the command manually. Here's what I have:
```
"C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.36.32532\bin\Hostx64\x64\cl.exe" /Fomodules\gdscript\gdscript_vm.windows.editor.x86_64.obj /c modules\gdscript\gdscript_vm.cpp /TP /std:c++17 /nologo /MT /Gd /GR /utf-8 /bigobj /O2 /W4 /wd4100 /wd4127 /wd4201 /wd4244 /wd4245 /wd4267 /wd4305 /wd4514 /wd4714 /wd4820 /EHsc /WX /DTOOLS_ENABLED /DDEBUG_ENABLED /DNDEBUG /DNO_EDITOR_SPLASH /DWINDOWS_ENABLED /DWASAPI_ENABLED /DWINMIDI_ENABLED /DTYPED_METHOD_BIND /DWIN32 /DMSVC /DWINVER=0x0601 /D_WIN32_WINNT=0x0601 /DNOMINMAX /D_WIN64 /DVULKAN_ENABLED /DGLES3_ENABLED /DMINIZIP_ENABLED /DBROTLI_ENABLED /DZSTD_STATIC_LINKING_ONLY /DUSE_VOLK /DVK_USE_PLATFORM_WIN32_KHR /DGLAD_ENABLED /DGLES_OVER_GL /DTESTS_ENABLED /Ithirdparty\libpng /Ithirdparty\glad /Ithirdparty\volk /Ithirdparty\vulkan /Ithirdparty\vulkan\include /Ithirdparty\zstd /Ithirdparty\zlib /Ithirdparty\brotli\include /Iplatform\windows /I.
```

(Actually I used `dev_mode=yes` instead of `verbose=yes`, so it also enabled warnings. This may negatively impact build time, I'll do another round of tests with just `scons p=windows target=editor verbose=yes`.)

The optimization flag is `/O2`, and can be replaced with other values for testing.

I did some comparisons with different optimization levels:

```
# `O2` (`optimize=speed` or `optimize=speed_trace`)
25m19.63s

# `O1` (`optimize=size`)
29m45.63s

# `Od` (`optimize=none` or `optimize=debug`)
2.93s
```

### Steps to reproduce

Compile `gdscript_vm.cpp` with `/O2` using MSVC, either with:

- `scons p=windows target=editor` (full engine build)
- `cl.exe /Fomodules\gdscript\gdscript_vm.windows.editor.x86_64.obj /c modules\gdscript\gdscript_vm.cpp /TP /std:c++17 /nologo /MT /Gd /GR /utf-8 /bigobj /O2 /DTOOLS_ENABLED /DDEBUG_ENABLED /DNDEBUG /DNO_EDITOR_SPLASH /DWINDOWS_ENABLED /DWASAPI_ENABLED /DWINMIDI_ENABLED /DTYPED_METHOD_BIND /DWIN32 /DMSVC /DWINVER=0x0601 /D_WIN32_WINNT=0x0601 /DNOMINMAX /D_WIN64 /DVULKAN_ENABLED /DGLES3_ENABLED /DMINIZIP_ENABLED /DBROTLI_ENABLED /DZSTD_STATIC_LINKING_ONLY /DUSE_VOLK /DVK_USE_PLATFORM_WIN32_KHR /DGLAD_ENABLED /DGLES_OVER_GL /DTESTS_ENABLED /Ithirdparty\libpng /Ithirdparty\glad /Ithirdparty\volk /Ithirdparty\vulkan /Ithirdparty\vulkan\include /Ithirdparty\zstd /Ithirdparty\zlib /Ithirdparty\brotli\include /Iplatform\windows /I.`
  * This compiles only `gdscript_vm.cpp` so it's faster / more focused, won't have you mislead thinking it's the linking step taking forever. You can change `/O2` for `/Od` to compare what it would be like with optimization off, and add/remove any flag which may have an impact on this build step.

### Minimal reproduction project

n/a

So my above analysis is slightly misleading, as I wasn't actually testing with `warnings=no`, but with removing these flags from the command line, which are all added in the `SConstruct` section which handles warnings for MSVC:
```
/W3 /w34458 /wd4100 /wd4127 /wd4201 /wd4244 /wd4245 /wd4267 /wd4305 /wd4514 /wd4714 /wd4820 /EHsc
```

And the flag which causes this massive build performance issue is actually not `/W3`, but `/EHsc`, which is set to silence this kind of warning emitted by thirdparty code:

```
C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.36.32532\include\vector(1137): error C2220: the following warning is treated as an error
C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.36.32532\include\vector(1241): note: see reference to function template instantiation 'void std::vector<T,glslang::pool_allocator<T>>::_Insert_counted_range<glslang::TArraySize*>(std::_Vector_const_iterator<std::_Vector_val<std::_Simple_types<_Ty>>>,_Iter,const unsigned __int64)' being compiled
        with
        [
            T=glslang::TArraySize,
            _Ty=glslang::TArraySize,
            _Iter=glslang::TArraySize *
        ]
D:\Dev\Godot\godot\thirdparty\glslang\glslang\Include\arrays.h(138): note: see reference to function template instantiation 'std::_Vector_iterator<std::_Vector_val<std::_Simple_types<_Ty>>> std::vector<_Ty,glslang::pool_allocator<T>>::insert<std::_Vector_iterator<std::_Vector_val<std::_Simple_types<_Ty>>>,0>(std::_Vector_const_iterator<std::_Vector_val<std::_Simple_types<_Ty>>>,_Iter,_Iter)' being compiled
        with
        [
            _Ty=glslang::TArraySize,
            T=glslang::TArraySize,
            _Iter=std::_Vector_iterator<std::_Vector_val<std::_Simple_types<glslang::TArraySize>>>
        ]
D:\Dev\Godot\godot\thirdparty\glslang\glslang\Include\arrays.h(138): note: see the first reference to 'std::vector<T,glslang::pool_allocator<T>>::insert' in 'glslang::TSmallArrayVector::push_back'
        with
        [
            T=glslang::TArraySize
        ]
C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.36.32532\include\vector(1137): warning C4530: C++ exception handler used, but unwind semantics are not enabled. Specify /EHsc
```

We can't specify `/EHsc` only for building the thirdparty code, because the warnings are also raised by our module code that makes use of those thirdparty headers.

After discussing with @bruvzg, I'm testing how it goes if we remove `/EHsc`, and simply silence `C4530`, with the following patch:
```diff
diff --git a/SConstruct b/SConstruct
index 6968967380..7c6a4ed45f 100644
--- a/SConstruct
+++ b/SConstruct
@@ -734,14 +734,12 @@ if selected_platform in platform_list:
                     "/wd4267",
                     "/wd4305",  # C4305 (truncation): double to float or real_t, too hard to avoid.
                     "/wd4514",  # C4514 (unreferenced inline function has been removed)
+                    "/wd4530",  # C4530 (C++ exception handler used without unwind semantics)
                     "/wd4714",  # C4714 (function marked as __forceinline not inlined)
                     "/wd4820",  # C4820 (padding added after construct)
                 ]
             )

-        # Set exception handling model to avoid warnings caused by Windows system headers.
-        env.Append(CCFLAGS=["/EHsc"])
-
         if env["werror"]:
             env.Append(CCFLAGS=["/WX"])
     else:  # GCC, Clang
diff --git a/methods.py b/methods.py
index 571a3f739e..99eedbf266 100644
--- a/methods.py
+++ b/methods.py
@@ -123,7 +123,7 @@ def disable_warnings(self):
         self["CCFLAGS"] = [x for x in self["CCFLAGS"] if not (x.startswith("/W") or x.startswith("/w"))]
         self["CFLAGS"] = [x for x in self["CFLAGS"] if not (x.startswith("/W") or x.startswith("/w"))]
         self["CXXFLAGS"] = [x for x in self["CXXFLAGS"] if not (x.startswith("/W") or x.startswith("/w"))]
-        self.AppendUnique(CCFLAGS=["/w"])
+        self.AppendUnique(CCFLAGS=["/w", "/wd4530"]) # Silence "C++ exception handler used without unwind semantics"
     else:
         self.AppendUnique(CCFLAGS=["-w"])
```

I don't know all the implications of the `/EH` flag, so this might need further review/discussion.
https://learn.microsoft.com/en-us/cpp/build/reference/eh-exception-handling-model?view=msvc-170

---

Test result:

```
# Clean Godot build with `/O2`, `warnings=all`, and `/EHsc`
35m   (didn't note down the precise number with seconds and too lazy to redo it)

# Clean Godot build with `/O2`, `warnings=all`, and `/wd4530`
15m57.171s
```

--- a/SConstruct
+++ b/SConstruct
@@ -193,6 +193,7 @@ opts.Add(BoolVariable("vulkan", "Enable the vulkan rendering driver", True))
 opts.Add(BoolVariable("opengl3", "Enable the OpenGL/GLES3 rendering driver", True))
 opts.Add(BoolVariable("openxr", "Enable the OpenXR driver", True))
 opts.Add(BoolVariable("use_volk", "Use the volk library to load the Vulkan loader dynamically", True))
+opts.Add(BoolVariable("disable_exceptions", "Force disabling exception handling code", True))
 opts.Add("custom_modules", "A list of comma-separated directory paths containing custom modules to build.", "")
 opts.Add(BoolVariable("custom_modules_recursive", "Detect custom modules recursively for each specified path.", True))
 
@@ -710,6 +711,16 @@ if selected_platform in platform_list:
             )
             Exit(255)
 
+    # Disable exception handling. Godot doesn't use exceptions anywhere, and this
+    # saves around 20% of binary size and very significant build time (GH-80513).
+    if env["disable_exceptions"]:
+        if env.msvc:
+            env.Append(CPPDEFINES=[("_HAS_EXCEPTIONS", 0)])
+        else:
+            env.Append(CCFLAGS=["-fno-exceptions"])
+    elif env.msvc:
+        env.Append(CCFLAGS=["/EHsc"])
+
     # Configure compiler warnings
     if env.msvc:  # MSVC
         if env["warnings"] == "no":
@@ -738,9 +749,6 @@ if selected_platform in platform_list:
                     "/wd4820",  # C4820 (padding added after construct)
                 ]
             )
-
-        # Set exception handling model to avoid warnings caused by Windows system headers.
-        env.Append(CCFLAGS=["/EHsc"])
 
         if env["werror"]:
             env.Append(CCFLAGS=["/WX"])

--- a/modules/denoise/SCsub
+++ b/modules/denoise/SCsub
@@ -109,6 +109,15 @@ env_oidn.AppendUnique(CPPDEFINES=["NDEBUG"])  # No assert() even in debug builds
 
 env_thirdparty = env_oidn.Clone()
 env_thirdparty.disable_warnings()
+
+if env["disable_exceptions"]:
+    # OIDN hard-requires exceptions, so we re-enable them here.
+    if env.msvc and ("_HAS_EXCEPTIONS", 0) in env_thirdparty["CPPDEFINES"]:
+        env_thirdparty["CPPDEFINES"].remove(("_HAS_EXCEPTIONS", 0))
+        env_thirdparty.AppendUnique(CCFLAGS=["/EHsc"])
+    elif not env.msvc and "-fno-exceptions" in env_thirdparty["CCFLAGS"]:
+        env_thirdparty["CCFLAGS"].remove("-fno-exceptions")
+
 env_thirdparty.add_source_files(thirdparty_obj, thirdparty_sources)
 env.modules_sources += thirdparty_obj
 

--- a/modules/openxr/SCsub
+++ b/modules/openxr/SCsub
@@ -52,10 +52,11 @@ if env["builtin_openxr"]:
 
     env_thirdparty = env_openxr.Clone()
     env_thirdparty.disable_warnings()
-    env_thirdparty.AppendUnique(CPPDEFINES=["DISABLE_STD_FILESYSTEM"])
 
-    if "-fno-exceptions" in env_thirdparty["CXXFLAGS"]:
-        env_thirdparty["CXXFLAGS"].remove("-fno-exceptions")
+    env_thirdparty.AppendUnique(CPPDEFINES=["DISABLE_STD_FILESYSTEM"])
+    if env["disable_exceptions"]:
+        env_thirdparty.AppendUnique(CPPDEFINES=["XRLOADER_DISABLE_EXCEPTION_HANDLING", ("JSON_USE_EXCEPTION", 0)])
+
     env_thirdparty.Append(CPPPATH=[thirdparty_dir + "/src/loader"])
 
     # add in external jsoncpp dependency

--- a/platform/android/detect.py
+++ b/platform/android/detect.py
@@ -170,10 +170,6 @@ def configure(env: "Environment"):
     env["RANLIB"] = compiler_path + "/llvm-ranlib"
     env["AS"] = compiler_path + "/clang"
 
-    # Disable exceptions on template builds
-    if not env.editor_build:
-        env.Append(CXXFLAGS=["-fno-exceptions"])
-
     env.Append(
         CCFLAGS=(
             "-fpic -ffunction-sections -funwind-tables -fstack-protector-strong -fvisibility=hidden -fno-strict-aliasing".split()

--- a/platform/ios/detect.py
+++ b/platform/ios/detect.py
@@ -30,7 +30,6 @@ def get_opts():
         ),
         ("IOS_SDK_PATH", "Path to the iOS SDK", ""),
         BoolVariable("ios_simulator", "Build for iOS Simulator", False),
-        BoolVariable("ios_exceptions", "Enable exceptions", False),
         ("ios_triple", "Triple for ios toolchain", ""),
     ]
 
@@ -137,11 +136,6 @@ def configure(env: "Environment"):
         )
         env.Append(ASFLAGS=["-arch", "arm64"])
         env.Append(CPPDEFINES=["NEED_LONG_INT"])
-
-    if env["ios_exceptions"]:
-        env.Append(CCFLAGS=["-fexceptions"])
-    else:
-        env.Append(CCFLAGS=["-fno-exceptions"])
 
     # Temp fix for ABS/MAX/MIN macros in iOS SDK blocking compilation
     env.Append(CCFLAGS=["-Wno-ambiguous-macro"])

--- a/platform/web/detect.py
+++ b/platform/web/detect.py
@@ -97,12 +97,9 @@ def configure(env: "Environment"):
     if env["use_assertions"]:
         env.Append(LINKFLAGS=["-s", "ASSERTIONS=1"])
 
-    if env.editor_build:
-        if env["initial_memory"] < 64:
-            print('Note: Forcing "initial_memory=64" as it is required for the web editor.')
-            env["initial_memory"] = 64
-    else:
-        env.Append(CPPFLAGS=["-fno-exceptions"])
+    if env.editor_build and env["initial_memory"] < 64:
+        print('Note: Forcing "initial_memory=64" as it is required for the web editor.')
+        env["initial_memory"] = 64
 
     env.Append(LINKFLAGS=["-s", "INITIAL_MEMORY=%sMB" % env["initial_memory"]])
 

--- a/tests/SCsub
+++ b/tests/SCsub
@@ -13,10 +13,8 @@ env_tests = env.Clone()
 if env_tests["platform"] == "windows":
     env_tests.Append(CPPDEFINES=[("DOCTEST_THREAD_LOCAL", "")])
 
-# Increase number of addressable sections in object files
-# due to doctest's heavy use of templates and macros.
-if env_tests.msvc:
-    env_tests.Append(CCFLAGS=["/bigobj"])
+if env["disable_exceptions"]:
+    env_tests.Append(CPPDEFINES=["DOCTEST_CONFIG_NO_EXCEPTIONS_BUT_WITH_ALL_ASSERTS"])
 
 env_tests.add_source_files(env.tests_sources, "*.cpp")
 

--- a/tests/core/io/test_image.h
+++ b/tests/core/io/test_image.h
@@ -262,9 +262,7 @@ TEST_CASE("[Image] Modifying pixels of an image") {
 
 		for (const Rect2i &rect : rects) {
 			Ref<Image> img = memnew(Image(img_width, img_height, false, Image::FORMAT_RGBA8));
-			CHECK_NOTHROW_MESSAGE(
-					img->fill_rect(rect, Color(1, 1, 1, 1)),
-					"fill_rect() shouldn't throw for any rect.");
+			img->fill_rect(rect, Color(1, 1, 1, 1));
 			for (int y = 0; y < img->get_height(); y++) {
 				for (int x = 0; x < img->get_width(); x++) {
 					if (rect.abs().has_point(Point2(x, y))) {
@@ -317,7 +315,7 @@ TEST_CASE("[Image] Modifying pixels of an image") {
 	// Pre-multiply Alpha then Convert from RGBA to L8, checking alpha
 	{
 		Ref<Image> gray_image = memnew(Image(3, 3, false, Image::FORMAT_RGBA8));
-		CHECK_NOTHROW_MESSAGE(gray_image->fill_rect(Rect2i(0, 0, 3, 3), Color(1, 1, 1, 0)), "fill_rect() shouldn't throw for any rect.");
+		gray_image->fill_rect(Rect2i(0, 0, 3, 3), Color(1, 1, 1, 0));
 		gray_image->set_pixel(1, 1, Color(1, 1, 1, 1));
 		gray_image->set_pixel(1, 2, Color(0.5, 0.5, 0.5, 0.5));
 		gray_image->set_pixel(2, 1, Color(0.25, 0.05, 0.5, 1.0));

--- a/tests/scene/test_viewport.h
+++ b/tests/scene/test_viewport.h
@@ -226,7 +226,7 @@ TEST_CASE("[SceneTree][Viewport] Controls and InputEvent handling") {
 
 	SUBCASE("[Viewport][GuiInputEvent] nullptr as argument doesn't lead to a crash.") {
 		ERR_PRINT_OFF;
-		CHECK_NOTHROW(root->push_input(nullptr));
+		root->push_input(nullptr);
 		ERR_PRINT_ON;
 	}
 

